File: packages/core/src/mcpact/contract/variables.py
"""Resolve ${{ steps.<name>.result.<path> }} references in contract values."""

from __future__ import annotations

import json
import re
from typing import Any

_VAR_RE = re.compile(r"\$\{\{\s*steps\.(\w+)\.result((?:\.\w+)*)\s*\}\}")


class VariableError(Exception):
    pass


def _walk_path(obj: Any, path: list[str]) -> Any:
    current = obj
    for segment in path:
        if isinstance(current, dict):
            if segment not in current:
                raise VariableError(f"path segment '{segment}' not found in {list(current.keys())}")
            current = current[segment]
        elif isinstance(current, list):
            try:
                current = current[int(segment)]
            except (ValueError, IndexError) as e:
                raise VariableError(f"invalid list index '{segment}'") from e
        else:
            raise VariableError(f"cannot traverse into {type(current).__name__} with '{segment}'")
    return current


def _extract_result(outcome_data: Any) -> Any:
    """Get a traversable result from a step's stored data.

    Prefers structured content (already a dict/list). Falls back to parsing
    text as JSON. Returns raw text if that also fails.
    """
    if isinstance(outcome_data, dict):
        structured = outcome_data.get("structured")
        if structured is not None:
            return structured
        text = outcome_data.get("text", "")
        if text:
            try:
                return json.loads(text)
            except (json.JSONDecodeError, TypeError):
                pass
        return text
    return outcome_data


def resolve_value(template: Any, context: dict[str, Any]) -> Any:
    """Resolve variable references in a single value.

    If the entire value is a single variable reference, preserves the resolved
    type (dict, list, int, etc.). Otherwise string-interpolates.
    """
    if not isinstance(template, str):
        if isinstance(template, dict):
            return {k: resolve_value(v, context) for k, v in template.items()}
        if isinstance(template, list):
            return [resolve_value(v, context) for v in template]
        return template

    match = _VAR_RE.fullmatch(template)
    if match:
        return _resolve_single(match, context)

    def _replacer(m: re.Match[str]) -> str:
        return str(_resolve_single(m, context))

    return _VAR_RE.sub(_replacer, template)


def _resolve_single(match: re.Match[str], context: dict[str, Any]) -> Any:
    step_name = match.group(1)
    path_str = match.group(2)

    if step_name not in context:
        raise VariableError(f"step '{step_name}' not found; available: {list(context.keys())}")

    result = _extract_result(context[step_name])
    if not path_str:
        return result

    path = [s for s in path_str.split(".") if s]
    return _walk_path(result, path)

File: packages/sdk-node/src/contract/variables.ts
const VAR_RE = /\$\{\{\s*steps\.(\w+)\.result((?:\.\w+)*)\s*\}\}/g;
const VAR_FULL_RE = /^\$\{\{\s*steps\.(\w+)\.result((?:\.\w+)*)\s*\}\}$/;

export class VariableError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "VariableError";
  }
}

function walkPath(obj: unknown, path: string[]): unknown {
  let current = obj;
  for (const segment of path) {
    if (current === null || current === undefined) {
      throw new VariableError(`cannot traverse into ${typeof current} with '${segment}'`);
    }
    if (typeof current === "object" && !Array.isArray(current)) {
      const rec = current as Record<string, unknown>;
      if (!(segment in rec)) {
        throw new VariableError(`path segment '${segment}' not found in [${Object.keys(rec).join(", ")}]`);
      }
      current = rec[segment];
    } else if (Array.isArray(current)) {
      const idx = Number(segment);
      if (Number.isNaN(idx) || idx < 0 || idx >= current.length) {
        throw new VariableError(`invalid list index '${segment}'`);
      }
      current = current[idx];
    } else {
      throw new VariableError(`cannot traverse into ${typeof current} with '${segment}'`);
    }
  }
  return current;
}

interface StepData {
  text?: string;
  structured?: unknown;
  is_error?: boolean;
}

function extractResult(data: unknown): unknown {
  if (typeof data === "object" && data !== null && !Array.isArray(data)) {
    const d = data as StepData;
    if (d.structured !== null && d.structured !== undefined) return d.structured;
    const text = d.text ?? "";
    if (text) {
      try {
        return JSON.parse(text);
      } catch {
        /* not JSON */
      }
    }
    return text;
  }
  return data;
}

function resolveSingle(stepName: string, pathStr: string, context: Record<string, unknown>): unknown {
  if (!(stepName in context)) {
    throw new VariableError(`step '${stepName}' not found; available: [${Object.keys(context).join(", ")}]`);
  }
  const result = extractResult(context[stepName]);
  if (!pathStr) return result;
  const path = pathStr.split(".").filter(Boolean);
  return walkPath(result, path);
}

export function resolveValue(template: unknown, context: Record<string, unknown>): unknown {
  if (typeof template === "string") {
    const full = VAR_FULL_RE.exec(template);
    if (full) {
      return resolveSingle(full[1]!, full[2]!, context);
    }
    return template.replace(VAR_RE, (_match, step: string, path: string) =>
      String(resolveSingle(step, path, context)),
    );
  }
  if (Array.isArray(template)) {
    return template.map((v) => resolveValue(v, context));
  }
  if (typeof template === "object" && template !== null) {
    const result: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(template)) {
      result[k] = resolveValue(v, context);
    }
    return result;
  }
  return template;
}
